gpt4all-backend/llmodel_c.cpp

```diff
@@ -51,6 +51,31 @@ void llmodel_llama_destroy(llmodel_model llama)
     delete wrapper;
 }
 
+llmodel_model llmodel_model_create(const char *model_path) {
+
+    uint32_t magic;
+    llmodel_model model;
+    FILE *f = fopen(model_path, "rb");
+    fread(&magic, sizeof(magic), 1, f);
+
+    if (magic == 0x67676d6c) { model = llmodel_gptj_create();  }
```

Comment:
This is cool. I read up on magic numbers for files but am not fully understanding how they are able to differentiate model types from different binary files. Just to improve my own understanding, can you go over how this method works in more detail - Is this a result of how you are creating the model files?

Reply:
I initially read it from niansa's [libjustlm](https://gitlab.com/niansa/libjustlm/-/blob/master/justlm.cpp) for GPTJ and thought: that's very clever!

What it does is that it reads the file in ifstream and then checks the size of the context. The magic numbers are just hexadecimals: 0x67676d6c(hex) = 1734831468(dec). All the different models have specific sizes in ggml. After reading the number for GPTJ I just printed them all and saw that there's certain sizes for MPT,LLaMa and GPTJ. So I just used those as checks and it seemed to work on my cli app.

What's really cool is that now that we check the type straight from the bin, you can **a)** have any names you want (not only ggml-whatever) and **b)** it does not accept any random binary which improves security a bit.

Reply:
In the future, will any issue come up if there's a model with different architecture but same sized context? Is that a scenario we need to consider?

Reply:
Yup. I think so.

If we happen to  get two models which have same magic number but are somehow incompatible, then yeah, we need some additional checks.

What you can do is to just simply put it in here:
```c++
    uint32_t magic;
    llmodel_model model;
    FILE *f = fopen(model_path, "rb");
    fread(&magic, sizeof(magic), 1, f);

    if (magic == 0xXXXXXXXX) {
    	if (some_additional_check(model_path) == true) {
    		model = llmodel_modelA_create();
    	} else {
    		model = llmodel_modelB_create();
    	}
    }
    if (magic == 0x67676d6c) { model = llmodel_gptj_create(); }
```

But what `some_additional_check` would be, I have no idea...

Reply:
Yeah. Good point!

I don't really know enough to have a proper answer but some guesses:

Since the magic numbers of GPTJ and MPT differ only by 1 (1734831468 vs 1734831469) I'm guessing that was a deliberate choice to differentiate the models. I don't really understand the ggml-alibi fork so maybe someone else could shed a light on this.

I think when we add a new model support, we bump the libllmodel version and update this function so that it will switch to the correct implementation. Maybe the `bool some_additional_check()` function then has to check the filename if the models have otherwise identical structure?

Right now if you had a hypothetical model with `magic=0x67676d6c` but its not a GPTJ model, and you call the undelying C `gptj_model_create()` and then `gptj_model_load(model)` with that model it would actually load **(!)** because inside `gptj_model_load` it does the magic check too to see if the model is valid GPTJ or not. Then it probably would just print nonsense tokens out.